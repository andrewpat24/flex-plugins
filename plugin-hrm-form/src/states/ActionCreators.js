import { HANDLE_BLUR,
         HANDLE_FOCUS
        } from './ActionTypes';
import { validateBeforeSubmit, 
         validateOnBlur,
         formIsValid } from './ValidationRules';

export const handleBlur = dispatch => (form, taskId) => () => {
  console.log("Received blur event");
  dispatch({
    type: HANDLE_BLUR,
    form: validateOnBlur(form),
    taskId
  });
}

export const handleFocus = dispatch => (taskId, parents, name) => {
  console.log(`handleFocus called with taskId = ${taskId}, parents = ${JSON.stringify(parents)}, name = ${name}`);
  dispatch({
    type: HANDLE_FOCUS,
    parents,
    name,
    taskId
  });
}

export const handleSubmit = dispatch => (form, handleCompleteTask) => (task) => {
  const newForm = validateBeforeSubmit(form);
  dispatch({
    type: HANDLE_BLUR, // probably need to rename this
    form: newForm,
    taskId: task.taskSid
  });
  if (formIsValid(newForm)) {
    handleCompleteTask(task.taskSid, task);
  } else {
    window.alert("There is a problem with your submission.  Please check the form for errors.");
  }
}