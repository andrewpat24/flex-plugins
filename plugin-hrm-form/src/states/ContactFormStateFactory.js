import cloneDeep from 'lodash/cloneDeep';
export const ValidationType = {
  REQUIRED: 'REQUIRED'
};
export const FieldType = {
  CALL_TYPE: 'CALL_TYPE',
  CHECKBOX: 'CHECKBOX',
  CHECKBOX_FIELD: 'CHECKBOX_FIELD',
  INTERMEDIATE: 'INTERMEDIATE',
  SELECT_SINGLE: 'SELECT_SINGLE',
  TAB: 'TAB',
  TEXT_BOX: 'TEXT_BOX',
  TEXT_INPUT: 'TEXT_INPUT'
};

const defaultFormDefinition = {
  callType: {
    type: FieldType.CALL_TYPE
  },
  callerInformation: {
    type: FieldType.TAB,
    name: {
      type: FieldType.INTERMEDIATE,
      firstName: {
        type: FieldType.TEXT_INPUT,
        validation: [ ValidationType.REQUIRED ]
      },
      lastName: {
        type: FieldType.TEXT_INPUT,
        validation: null
      }
    },
    relationshipToChild: {
      type: FieldType.SELECT_SINGLE,
      validation: null
    },
    gender: {
      type: FieldType.SELECT_SINGLE,
      validation: [ ValidationType.REQUIRED ]
    },
    age: {
      type: FieldType.SELECT_SINGLE,
      validation: [ ValidationType.REQUIRED ]
    }
  },
  caseInformation: {
    type: FieldType.TAB,
    categories: {
      type: FieldType.CHECKBOX_FIELD,
      validation: [ ValidationType.REQUIRED ],
      category1: {
        sub1: false,
        sub2: false,
        sub3: false,
        sub4: false,
        sub5: false,
        sub6: false,
      },
      category2: {
        sub1: false,
        sub2: false,
        sub3: false,
        sub4: false,
        sub5: false,
        sub6: false,
      }
    },
    keepConfidential: {
      type: FieldType.CHECKBOX,
      value: true
    },
    okForCaseWorkerToCall: {
      type: FieldType.CHECKBOX,
      value: false
    }
  }
};

export const generateInitialFormState = (formDefinition = defaultFormDefinition) => {
  let initialState = {};
  Object.keys(formDefinition).filter(key => (key !== 'type')).forEach(key => {
    switch (formDefinition[key].type) {
      case FieldType.CALL_TYPE:
        initialState[key] = {
          ...formDefinition[key],
          value: ''
        };
        break;
      case FieldType.CHECKBOX:
        initialState[key] = {
          value: false,  // set default of false if not overridden
          ...formDefinition[key]
        };
        break;
      case FieldType.CHECKBOX_FIELD:
        initialState[key] = cloneDeep(formDefinition[key]);
        initialState[key].error = null;
        break;
      case FieldType.INTERMEDIATE:
      case FieldType.TAB:
        initialState[key] = {
          ...generateInitialFormState(formDefinition[key]),
          type: formDefinition[key].type
        };
        break;
      case FieldType.SELECT_SINGLE:
      case FieldType.TEXT_BOX:
      case FieldType.TEXT_INPUT:
        initialState[key] = {
          value: '',  // set default of empty if not overridden
          ...formDefinition[key],
          touched: false,
          error: null
        };
        break;
      default:
        throw new Error(`Unknown FieldType ${formDefinition[key].type} for key ${key} in ${JSON.stringify(formDefinition)}`);
    }
  });
  return initialState;
}
