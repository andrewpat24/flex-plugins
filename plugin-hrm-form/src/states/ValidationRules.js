import { callTypes } from './DomainConstants';
import cloneDeep from 'lodash/cloneDeep';

export function validateFormBeforeSubmit(form) {
  // return isStandAloneCallType(form.callType);
  return true; // need to add more later
}

// Questionable whether we should export this
export function isStandAloneCallType(callType) {
  return (callType !== callTypes.caller &&
          callType !== callTypes.child);
}

// returns a new version of the form that updates the error messages
// should this actually check for whether anything has changed?
// not planning to do that yet, but could add it as an optimization
export function validateOnBlur(form) {
  let newForm = cloneDeep(form);
  if (newForm.callerInformation.name.firstName.touched &&
      !newForm.callerInformation.name.firstName.value) {
        newForm.callerInformation.name.firstName.error = "This field is required";
  } else {
    newForm.callerInformation.name.firstName.error = null;
  }
  return newForm;
}

// walk a form tree looking for non-null error values
export function formIsValid(form) {
  if ('error' in form && form.error !== null) {
    return false;
  }
  for (const key of Object.keys(form)) {
    if (typeof form[key] === 'object' 
        && form[key] !== null
        && !formIsValid(form[key])) {
      return false;
    }
  }
  return true;
};