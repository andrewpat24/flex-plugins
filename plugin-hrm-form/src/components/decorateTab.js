import React from 'react';
import { Tab } from '@material-ui/core';
import ErrorIcon from '@material-ui/icons/Error';
import { formIsValid } from '../states/ValidationRules';

const decorateTab = (label, formRoot) => {
  if (formIsValid(formRoot)) {
    return (
      <Tab label={label} />
    );  
  } else {
    return (
      <Tab label={<><span style={{verticalAlign: 'middle'}}><ErrorIcon fontSize="inherit" /></span> {label} </>} />
    );
  }
}

export default decorateTab;