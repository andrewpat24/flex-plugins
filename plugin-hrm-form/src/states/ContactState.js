import { saveToHrm } from '../components/HrmFormController';
import isEqual from 'lodash/isEqual';
import { generateInitialFormState } from './ContactFormStateFactory';

import { HANDLE_BLUR,
         HANDLE_CHANGE,
         HANDLE_CHANGE_NEW,
         HANDLE_FOCUS,
         INITIALIZE_CONTACT_STATE,
         REMOVE_CONTACT_STATE,
         SAVE_CONTACT_STATE
} from './ActionTypes';

const initialState = {
  tasks: { }
};

const taskInitialStateFactory = () => {
    return {
      callType: '',
      internal: {
        tab: 0
      },
      callerInformation: {
        name: {
          firstName: {
            value: '',
            touched: false,
            error: null
          },
          lastName: ''
        },
        relationshipToChild: '',
        gender: '',
        age: '',
        language:'',
        nationality: '',
        ethnicity: '',
        location: {
          streetAddress: '',
          city: '',
          stateOrCounty: '',
          postalCode: '',
          phone1: '',
          phone2: ''
        }
      },
      childInformation: {
        name: {
          firstName: '',
          lastName: ''
        },
        gender: {
          value: '',
          touched: false,
          error: null
        },
        age: '',
        language:'',
        nationality: '',
        ethnicity: '',
        school: {
          name: '',
          gradeLevel: ''
        },
        location: {
          streetAddress: '',
          city: '',
          stateOrCounty: '',
          postalCode: '',
          phone1: '',
          phone2: ''
        },
        refugee: false,
        disabledOrSpecialNeeds: false,
        hiv: false
      },
      caseInformation: {
        categories: {
          category1: {
            sub1: false,
            sub2: false,
            sub3: false,
            sub4: false,
            sub5: false,
            sub6: false,
          },
          category2: {
            sub1: false,
            sub2: false,
            sub3: false,
            sub4: false,
            sub5: false,
            sub6: false,
          },
          category3: {
            sub1: false,
            sub2: false,
            sub3: false,
            sub4: false,
            sub5: false,
            sub6: false,
          },
          category4: {
            sub1: false,
            sub2: false,
            sub3: false,
            sub4: false,
            sub5: false,
            sub6: false,
          },
          category5: {
            sub1: false,
            sub2: false,
            sub3: false,
            sub4: false,
            sub5: false,
            sub6: false,
          },
          category6: {
            sub1: false,
            sub2: false,
            sub3: false,
            sub4: false,
            sub5: false,
            sub6: false,
          }
        },
        callSummary: '',
        referredTo: '',
        status: 'In Progress',
        keepConfidential: true,
        okForCaseWorkerToCall: false,
        howDidTheChildHearAboutUs: '',
        didYouDiscussRightsWithTheChild: false,
        didTheChildFeelWeSolvedTheirProblem: false,
        wouldTheChildRecommendUsToAFriend: false
      }
    };
};

export class Actions {
  // static handleChange = (taskId, parents, e) => ({type: HANDLE_CHANGE, name: e.currentTarget.name, value: e.currentTarget.value, taskId: taskId, parents: parents});
  static handleChange = function(taskId, parents, e) {
    console.log(e);
    return {type: HANDLE_CHANGE, name: e.target.name || e.currentTarget.name, value: e.target.value || e.currentTarget.value, taskId: taskId, parents: parents};
  };
  // This makes me so sad too
  // static handleCheckbox = (taskId, parents, name, value) => ({type: HANDLE_CHANGE, name: name, taskId: taskId, value: value, parents: parents});
  static handleCheckbox = (taskId, parents, name, value) => ({type: HANDLE_CHANGE, name: name, taskId: taskId, value: value, parents: parents});
  // There has to be a better way to do this rather than a one-off, but MUI does not make it easy
  // static handleCallTypeButtonClick = (taskId, value, e) => ({type: HANDLE_CALLTYPE_BUTTON_CLICK, taskId: taskId, value: value});
  static handleCallTypeButtonClick = (taskId, value, e) => ({type: HANDLE_CHANGE_NEW, name: 'callType', taskId: taskId, value: value, parents: []});
  static initializeContactState = (taskId) => ({type: INITIALIZE_CONTACT_STATE, taskId: taskId});
  // I'm really not sure if this should live here, but it seems like we need to come through the store
  static saveContactState = (task, abortFunction, hrmBaseUrl) => ({type: SAVE_CONTACT_STATE, task, abortFunction, hrmBaseUrl});
  static removeContactState = (taskId) => ({type: REMOVE_CONTACT_STATE, taskId: taskId});
}

// Will replace the below when we move over to field objects
function editNestedFieldNew(original, parents, name, change) {
  if (parents.length === 0) {
    return {
      ...original,
      [name] : {
        ...original[name],
        ...change
      }
    };
  }
  return {
    ...original,
    [parents[0]]: editNestedFieldNew(original[parents[0]], parents.slice(1), name, change)
  }
}

/// TEST THIS AND MAKE IT GENERALLLLLLLLL
// VisibleForTesting
export function editNestedField(original, parents, name, value) {
  // This is horrible, I promise I won't leave it for long
  if (name === 'firstName' &&
      isEqual(parents, ['callerInformation', 'name'])) {
    return {
      ...original,
      callerInformation: {
        ...original.callerInformation,
        name: {
          ...original.callerInformation.name,
          firstName: {
            ...original.callerInformation.name.firstName,
            value
          }
        }
      }
    }
  }
  if (name === 'gender' &&
      isEqual(parents, ['childInformation'])) {
    return {
      ...original,
      childInformation: {
        ...original.childInformation,
        gender: {
          ...original.callerInformation.gender,
          value
        }
      }
    }
  }
  if (parents.length === 0) {
    return { 
      ...original,
      [name] : value
    };
  }
  return {
    ...original,
    [parents[0]]: editNestedField(original[parents[0]], parents.slice(1), name, value)
  }
}

export function reduce(state = initialState, action) {
  switch (action.type) {
    case HANDLE_BLUR: {
      return {
        ...state,
        tasks: {
          ...state.tasks,
          [action.taskId]: action.form
        }
      };
    }

    case HANDLE_CHANGE_NEW: {
      let currentForm;
      if (state.tasks[action.taskId]) {
        currentForm = state.tasks[action.taskId];
      } else {
        // currentForm = taskInitialStateFactory();
        currentForm = generateInitialFormState();
        console.log("Had to recreate state for taskId " + action.taskId);
      }
      return {
        ...state,
        tasks: {
          ...state.tasks,
          [action.taskId]: editNestedFieldNew(currentForm,
                                              action.parents,
                                              action.name,
                                              {value: action.value})
        }
      };
    }

    case HANDLE_FOCUS: {
      let currentForm;
      if (state.tasks[action.taskId]) {
        currentForm = state.tasks[action.taskId];
      } else {
        // currentForm = taskInitialStateFactory();
        currentForm = generateInitialFormState();
        console.log("Had to recreate state for taskId " + action.taskId);
      }
      return {
        ...state,
        tasks: {
          ...state.tasks,
          [action.taskId]: editNestedFieldNew(currentForm,
                                              action.parents,
                                              action.name,
                                              {touched: true})
        }
      };
    }

    case HANDLE_CHANGE: {
      console.log("!!!!!!!!!!!HANDLE CHANGE: action.name = " + action.name + ", action.value = " + action.value + ", task = " + action.taskId + ", parents: " + action.parents);
      // let updatedContactForm = state.tasks[action.taskId];
      // updatedContactForm = {
      //   ...updatedContactForm,
      //   [action.name]: action.value
      // };
      // we could probably replace the below if/else by having the first argument to editNestedField be
      //  state.tasks[action.taskId] || taskInitialStateFactory()
      // but I want to be more explicit and log it.  Redux gets purged if there's a refresh and that's messy
      let currentForm;
      if (state.tasks[action.taskId]) {
        currentForm = state.tasks[action.taskId];
      } else {
        // currentForm = taskInitialStateFactory();
        currentForm = generateInitialFormState();
        console.log("Had to recreate state for taskId " + action.taskId);
      }
      return {
        ...state,
        tasks: {
          ...state.tasks,
          // [action.taskId]: editNestedField(currentForm,
          //                                  action.parents,
          //                                  action.name,
          //                                  action.value)
          [action.taskId]: editNestedFieldNew(currentForm,
                                              action.parents,
                                              action.name,
                                              {value: action.value})
        }
      };
    }

    case INITIALIZE_CONTACT_STATE: {
      console.log("!!!!!!!!!CREATING NEW ENTRY FOR " + action.taskId);
      return {
        ...state,
        tasks: {
          ...state.tasks,
          [action.taskId]: generateInitialFormState() //taskInitialStateFactory()
        }
      }
    }

    case SAVE_CONTACT_STATE: {
      // TODO(nick): Make this a Promise instead?
      saveToHrm(action.task, state.tasks[action.task.taskSid], action.abortFunction, action.hrmBaseUrl);
      return state;
    }

    case REMOVE_CONTACT_STATE: {
      console.log("!!!!!!!!!DELETING ENTRY FOR " + action.taskId);
      return {
        ...state,
        tasks: {
          ...state.tasks,
          [action.taskId]: undefined
        }
      }
    }

    default:
      return state;
  }
}
