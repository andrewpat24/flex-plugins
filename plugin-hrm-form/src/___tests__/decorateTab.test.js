jest.mock('@material-ui/core', () => {
  return {
    Tab: 'Tab'
  };
});
jest.mock('../states/ValidationRules', () => {
  return {
    formIsValid: jest.fn()
  };
});
jest.mock('@material-ui/icons/Error', () => 'ErrorIcon');
import decorateTab from '../components/decorateTab';
import { Tab } from '@material-ui/core';
import ErrorIcon from '@material-ui/icons/Error';
import { formIsValid } from '../states/ValidationRules';
import renderer from 'react-test-renderer';

test('decorateTab when valid', () => {
  formIsValid.mockReturnValueOnce(true);
  const component = renderer.create(
    decorateTab("My Label", {})
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('decorateTab when invalid', () => {
  formIsValid.mockReturnValueOnce(false);
  const component = renderer.create(
    decorateTab("My Label", {})
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});