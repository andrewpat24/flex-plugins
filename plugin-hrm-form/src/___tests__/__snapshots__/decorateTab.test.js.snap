// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorateTab when invalid 1`] = `
<Tab
  label={
    <React.Fragment>
      <span
        style={
          Object {
            "verticalAlign": "middle",
          }
        }
      >
        <ErrorIcon
          fontSize="inherit"
          style={
            Object {
              "color": "red",
            }
          }
        />
      </span>
       
      My Label
       
    </React.Fragment>
  }
/>
`;

exports[`decorateTab when valid 1`] = `
<Tab
  label="My Label"
/>
`;
